monai/data/wsi_reader.py

```diff
@@ -107,37 +133,105 @@ def set_device(self, device):
+            power: the objective power.
+
+        """
+
+        # Try instance parameters if no resolution is provided
+        if mpp is None and power is None and level is None:
+            mpp = self.mpp
+            power = self.power
+            level = self.level
+
+        # Ensure that at most one resolution parameter is provided.
+        resolution = [val[0] for val in [("level", level), ("mpp", mpp), ("power", power)] if val[1] is not None]
+        if len(resolution) > 1:
+            raise ValueError(f"Only one of `level`, `mpp`, or `power` should be provided. {resolution} are provided.")
+
+        n_levels = self.get_level_count(wsi)
+
+        if mpp is not None:
+            mpp_: tuple[float, float] = ensure_tuple_rep(mpp, 2)  # type: ignore
+            available_mpps = [self.get_mpp(wsi, level) for level in range(n_levels)]
+            if mpp_ in available_mpps:
+                valid_mpp = mpp_
+            else:
+                valid_mpp = min(available_mpps, key=lambda x: abs(x[0] - mpp_[0]) + abs(x[1] - mpp_[1]))
+                for i in range(2):
+                    if abs(valid_mpp[i] - mpp_[i]) > self.mpp_atol + self.mpp_rtol * abs(mpp_[i]):
+                        raise ValueError(
+                            f"The requested mpp {mpp_} does not exist in this whole slide image "
+                            f"(with mpp_rtol={self.mpp_rtol} and mpp_atol={self.mpp_atol}). "
+                            f"Here is the list of available mpps: {available_mpps}. "
+                            f"The closest matching available mpp is {valid_mpp}."
+                            "Please consider changing the tolerances or use another mpp."
+                        )
+            level = available_mpps.index(valid_mpp)
+
+        elif power is not None:
+            available_powers = [self.get_power(wsi, level) for level in range(n_levels)]
+            if power in available_powers:
+                valid_power = power
+            else:
+                valid_power = min(available_powers, key=lambda x: abs(x - power))  # type: ignore
+                if abs(valid_power - power) > self.power_atol + self.power_rtol * abs(power):
+                    raise ValueError(
+                        f"The requested power ({power}) does not exist in this whole slide image "
+                        f"(with power_rtol={self.power_rtol} and power_atol={self.power_atol})."
+                        f"Here is the list of available objective powers: {available_powers}. "
+                        f" The closest matching available power is {valid_power}."
+                        "Please consider changing the tolerances or use another power."
+                    )
+            level = available_powers.index(valid_power)
+        else:
+            if level is None:
+                # Set the default value if no resolution parameter is provided.
+                level = 0
+            if level >= n_levels:
```

Comment:
these are some duplications here, could you make it a utility function as well? also please use `np.isclose` or `np.allclose` instead of a customized implementation

Reply:
The error messages here are very specific, so I don't think that we can gain anything by creating a utility function. Is there other places that we might need such logic?

Reply:
at least locally it's just replacing 'power' vs 'mpp' and all the rest of the logic are the same. please simplify them here, if it's not a utility function, at least these could be merged

Reply:
Sure, I merged them.